README
IO-Pipe-Producer version 0.01
=============================


                                 Producer.pm
                                 Version 1.6

  What is it?
  -----------

  Producer.pm is a module that provides methods to fork off a subroutine or
  system call and return handles on the standard output (STDOUT and STDERR).
  If you have (for example) a subroutine that processes a very large text file
  and performs a task on each line, but you need to perform further processing
  with either other subroutines or in main, normally you would have to wait
  until the subroutine returns to get its output (either by returning it or
  opening a temporary file that it produced) and continue processing.  If the
  subroutine prints its output to STDOUT (and STDERR) or you can edit it to do
  so, you can call it using a producer so that you can use the returned handle
  to continuously process each line as it's "produced".  You can chain
  subroutines together like this by having your subroutine itself create a
  Producer.  This is similar to using open() to run a system call, except that
  with this module, you can get a handle on STDERR and use it with subroutines
  as well.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires these other modules and libraries:

  IO::Pipe
  Carp

COPYRIGHT AND LICENCE

This software and ancillary information (herein called "SOFTWARE") called
Producer.pm is made available under the terms described here.  The
SOFTWARE has been approved for release with associated LA-CC number
LA-CC-05-060.

Unless otherwise indicated, this software has been authored by an employee or
employees of the University of California, operator of the Los Alamos National
Laboratory under Contract No. W-7405-ENG-36 with the U.S. Department of
Energy.  The U.S. government has rights to use, reproduce, and distribute this
SOFTWARE.  The public may copy, distribute, prepare derivative works and
publicly display this SOFTWARE without charge, provided that this notice and
any statement of authorship are reproduced on all copies.  Neither the
government nor the university makes any warranty, express or implied, or
assumes any liability or responsibility for the use of this SOFTWARE.

If SOFTWARE is modified to produce derivative works, such modified SOFTWARE
should be clearly marked, so as not to confuse it with the version available
from LANL.



Changes
Revision history for Perl extension IO::Pipe::Producer.

1.0  2003/03/07 by robleach

    Initial version.

1.1  2005/04/11 by robleach

    I added the ability to return handles on both standard output and standard
    error (when called in array context).  I also added a subroutine to do the
    same thing with a system call (as opposed to a subroutine call).

1.2  2005/04/11 by robleach

    I added the ability to return handles on both standard output and standard
    error (when called in array context).  I also added a subroutine to do the
    same thing with a system call (as opposed to a subroutine call).

1.3  2005/04/11 by robleach

    This is a modular version of producer_fork.pl.  I changed the method names,
    made a constructor, and changed the documentation.  It now ISA IO::Pipe and
    all the file handles returned are IO::Pipe::Producer instances.

1.4  2005/05/04 by robleach

    I incorporated Carp.pm, pod documentation, created $Producer::VERSION and
    $Producer::errstr variables, and edited/updated some comments.

1.5  2010/09/01 by robleach

    Updated LANL license info.

1.6  Fri Nov 16 10:51:29 2012 by robleach

    Standardized with Makemaker and updated perldoc documentation.
    Created by h2xs 1.23 with options
        -AX --skip-exporter --use-new-tests -n IO::Pipe::Producer

1.7  2014/07/11 by robleach

    Changed documentation to inform user that they must close the returned file
    handles and added a check for finished unclosed file handles.  Closes them
    if there are over 200.

1.8  2014/07/15 by robleach

    Based on test results, I reduced the maximum number of allowed unclosed
    file handles to 100, as it seemed the most any system tested could handle
    was 110.

1.9  2014/07/18 by robleach

    Added a call to tell before calling eof when checking for expired file
    handles so as to avoid blocking IO situations.

2.0  2014/07/22 by robleach

    Based on CPAN tester test results, removed file handle tracking.


test.pl
#!/usr/bin/perl

use strict;

#Print the number of tests in a format that Module::Build understands
print("1..6\n");

#Initialize vars needed by IO::Pipe::Producer
my $subroutine_reference = \&test;
my @subroutine_parameters = ('testing',1,2,3);

#Keep track of how many tests failed
my $num_failed = 0;

#use the module I want to test
use IO::Pipe::Producer;

##
## Test 1
##

#Test a scalar context call to getSubroutineProducer
my $obj = new IO::Pipe::Producer();
my $stdout_fh =
  $obj->getSubroutineProducer($subroutine_reference,@subroutine_parameters);
my @output = map {chomp;$_} <$stdout_fh>;

if(equaleq(\@subroutine_parameters,\@output))
  {print("ok 1\n")}
else
  {
    $num_failed++;
    print("not ok 1\n");
  }

##
## Test 2
##

#Test a list context call to getSubroutineProducer
$obj = new IO::Pipe::Producer();
my($stderr_fh);
($stdout_fh,$stderr_fh) =
  $obj->getSubroutineProducer($subroutine_reference,@subroutine_parameters);
@output          = map {chomp;$_} <$stdout_fh>;
my @error_output = map {chomp;$_} <$stderr_fh>;

if(equaleq(\@subroutine_parameters,\@output) &&
   equaleq(\@subroutine_parameters,\@error_output))
  {print("ok 2\n")}
else
  {
    $num_failed++;
    print("not ok 2\n");
  }

##
## Test 3
##

#Test a scalar context call to new that returns an STDOUT handle
$stdout_fh = new IO::Pipe::Producer($subroutine_reference,
				    @subroutine_parameters);
@output = map {chomp;$_} <$stdout_fh>;

if(equaleq(\@subroutine_parameters,\@output))
  {print("ok 3\n")}
else
  {
    $num_failed++;
    print("not ok 3\n");
  }

##
## Test 4
##

#Test a list context call to new that returns STDOUT & STDERR handles
($stdout_fh,$stderr_fh) =
  new IO::Pipe::Producer($subroutine_reference,@subroutine_parameters);
@output       = map {chomp;$_} <$stdout_fh>;
@error_output = map {chomp;$_} <$stderr_fh>;

if(equaleq(\@subroutine_parameters,\@output) &&
   equaleq(\@subroutine_parameters,\@error_output))
  {print("ok 4\n")}
else
  {
    $num_failed++;
    print("not ok 4\n");
  }

##
## Test 5
##

#Test a scalar context call to getSystemProducer
$obj = new IO::Pipe::Producer();
$stdout_fh = $obj->getSystemProducer("echo \"Hello World!\"");
@output = map {chomp;$_} <$stdout_fh>;

if("Hello World!" eq $output[0])
  {print("ok 5\n")}
else
  {
    print STDERR ("[@output] ne [Hello World!]\n");
    $num_failed++;
    print("not ok 5\n");
  }

##
## Test 6
##

#Test a list context call to getSystemProducer
$obj = new IO::Pipe::Producer();
($stdout_fh,$stderr_fh) =
  $obj->getSystemProducer("perl -e 'print \"Hello World!\";" .
			  "print STDERR join(\"\\n\",(" .
			  join(',',@subroutine_parameters) .
			  "))'");
@output       = map {chomp;$_} <$stdout_fh>;
@error_output = map {chomp;$_} <$stderr_fh>;

if("Hello World!" eq $output[0] &&
   equaleq(\@subroutine_parameters,\@error_output))
  {print("ok 6\n")}
else
  {
    print STDERR (('Hello World!' ne $output[0] ?
		   "[@output] ne [Hello World!] in STDOUT" : ''),
		  (equaleq(\@subroutine_parameters,\@error_output) ?
		   '':"[@error_output] ne [@subroutine_parameters] in STDERR"),
		  "\n");
    $num_failed++;
    print("not ok 6\n");
  }

#Exit with the number of failed tests
exit($num_failed);



#Subroutine that will be sent to the method calls in IO::Pipe::Producer
#It simply prints the parameters sent in separated by new lines
sub test
  {
    print(join("\n",@_));
    print STDERR (join("\n",@_));
  }

#The chomp'd output of the test subroutine will be tested against the list of
#arguments sent in using this subroutine which basically does a string compare
#of array elements (and a size sompare of the two arrays sent in)
sub equaleq
  {
    my $ary1 = $_[0];
    my $ary2 = $_[1];

    #If the arrays aren't the same size, issue an error and return 0 (false)
    if(scalar(@$ary1) != scalar(@$ary2))
      {
	print STDERR ("Arrays are not the same size.  The first array has ",
		      scalar(@$ary1),
		      " elements and the second array has ",
		      scalar(@$ary2),
		      " elements.\n");
	return(0);
      }

    #If any of the (assumed) scalar elements don't match, issue an error and
    #return 0 (false)
    foreach my $index (0..$#{$ary1})
      {
	if($ary1->[$index] ne $ary2->[$index])
	  {
	    print STDERR ("Elements[$index] are not the same: ",
			  "[$ary1->[$index]] ne [$ary2->[$index]].");
	    return(0);
	  }
      }

    #Return 1 (true)
    return(1);
  }



lib/IO/Pipe/Producer.pm
package IO::Pipe::Producer;

use 5.012003;
use strict;
use warnings;
use Carp;

our @ISA = qw(IO::Pipe);
use base qw(IO::Pipe);

our $VERSION = '2.0';

#NOTICE
#
#This software and ancillary information (herein called "SOFTWARE") called
#Producer.pm is made available under the terms described here.  The
#SOFTWARE has been approved for release with associated LA-CC number
#LA-CC-05-060.
#
#Unless otherwise indicated, this software has been authored by an employee or
#employees of the University of California, operator of the Los Alamos National
#Laboratory under Contract No. W-7405-ENG-36 with the U.S. Department of
#Energy.  The U.S. government has rights to use, reproduce, and distribute this
#SOFTWARE.  The public may copy, distribute, prepare derivative works and
#publicly display this SOFTWARE without charge, provided that this notice and
#any statement of authorship are reproduced on all copies.  Neither the
#government nor the university makes any warranty, express or implied, or
#assumes any liability or responsibility for the use of this SOFTWARE.
#
#If SOFTWARE is modified to produce derivative works, such modified SOFTWARE
#should be clearly marked, so as not to confuse it with the version available
#from LANL.


#Constructor

sub new
  {
    #Get the class name
    my $class = shift(@_);
    #Instantiate an instance of the super class
    my $self = $class->SUPER::new();
    #Bless the instantiation into this class so we can call our own methods
    bless($self,$class);
    #If a subroutine call was supplied
    if(scalar(@_))
      {
	#Declare file handles for STDOUT and STDERR
	my($fh,$eh);
	#If new was called in list context
	if(wantarray)
	  {
	    #Fill the handles with the outputs from the subroutine
	    ($fh,$eh) = $self->getSubroutineProducer(@_);
	    #Return blessed referents to the file handles
	    return(bless($fh,$class),bless($eh,$class));
	  }
	#Fill the STDOUT handle with the output from the subroutine
	$fh = $self->getSubroutineProducer(@_);
	#Return blessed referent to the STDOUT handle
	return(bless($fh,$class));
      }
    #Return a blessed referent of the object hash
    if(wantarray)
      {return($self,bless($class->SUPER::new(),$class))}
    return($self);
  }



#This method is also a constructor
sub getSubroutineProducer
  {
    #Read in subroutine reference
    my $self         = shift;
    my $producer_sub = shift;
    my @params       = @_;
    my($pid,$error);

    if(!defined($producer_sub) || ref($producer_sub) ne 'CODE')
      {
	$error = "ERROR:Producer.pm:getSubroutineProducer:A referenced " .
	  "subroutine is required as the first argument to " .
	    "getSubroutineProducer.";
	$Producer::errstr = $error;
	carp($error);
	return(undef);
      }

    #Create a pipe
    my $stdout_pipe = $self->SUPER::new();
    my($stderr_pipe);
    $stderr_pipe = $self->SUPER::new() if(wantarray);

    #Fork off the Producer
    if(defined($pid = fork()))
      {
	if($pid)
	  {
	    ##
	    ## Parent
	    ##

	    #Create a read file handle
	    $stdout_pipe->reader();
	    $stderr_pipe->reader() if(wantarray);

	    #Return the read file handle to the consumer
	    if(wantarray)
	      {return(bless($stdout_pipe,ref($self)),
		      bless($stderr_pipe,ref($self)))}
	    return(bless($stdout_pipe,ref($self)));
	  }
	else
	  {
	    ##
	    ## Child
	    ##

	    #Create a write file handle for the Producer
	    $stdout_pipe->writer();
	    $stdout_pipe->autoflush;
	    $stderr_pipe->writer()  if(defined($stderr_pipe));
	    $stderr_pipe->autoflush if(defined($stderr_pipe));

	    #Redirect standard outputs to the pipes or kill the child
	    if(!open(STDOUT,">&",\${$stdout_pipe}))
	      {
		$error = "ERROR:Producer.pm:getSubroutineProducer:Can't " .
		  "redirect stdout to pipe: [" .
		    select($stdout_pipe) .
		      "]. $!";
		$Producer::errstr = $error;
		croak($error);
	      }
	    elsif(defined($stderr_pipe) && !open(STDERR,">&",\${$stderr_pipe}))
	      {
		$error = "ERROR:Producer.pm:getSubroutineProducer:Can't " .
		  "redirect stderr to pipe: [" .
		    select($stderr_pipe) .
		      "]. $!";
		$Producer::errstr = $error;
		croak($error);
	      }

	    #Call the subroutine passed in (ignore it's return value)
	    $producer_sub->(@params);

	    #Close the writer pipes
	    close($stdout_pipe);
	    close($stderr_pipe) if(defined($stderr_pipe));

	    #Successfully exiting the child process
	    exit(0);
	  }
      }
    else
      {
	$error = "ERROR:Producer.pm:getSubroutineProducer:fork() didn't work!";
	$Producer::errstr = $error;
	carp($error);
	return(undef);
      }
  }


sub getSystemProducer
  {
    my $self = shift;
    return($self->getSubroutineProducer(sub {system(@_)},@_));
  }


1;
__END__

=head1 NAME

IO::Pipe::Producer - Perl extension for IO::Pipe

=head1 SYNOPSIS
 
  #Example 1 (Call a subroutine & grab its standard output):
 
  use IO::Pipe::Producer;
  $obj = new IO::Pipe::Producer();
  $stdout_file_handle =
    $obj->getSubroutineProducer(\&mysub,
                                @mysub_params);
  while(<$stdout_file_handle>)
    {print}
  close($stdout_file_handle);
 
  #Example 2 (Call a subroutine & grab its standard output & standard error):
 
  ($stdout_file_handle,$stderr_file_handle) =
    $obj->getSubroutineProducer(\&mysub,
                                @mysub_params);
 
  #It is recommended to use IO::Select when reading more than 1 file handle:
  use IO::Select;
  my $sel = new IO::Select;
  $sel->add($stdout_file_handle,$stderr_file_handle);
  while(my @fhs = $sel->can_read())
    {
      foreach my $fh (@fhs)
        {
          my $line = <$fh>;
          unless(defined($line))
            {
              $sel->remove($fh);
              close($fh);
              next;
            }
          if($fh == $stdout_file_handle)
            {$messages .= $line}
          elsif($fh == $stderr)
            {$errors .= $line}
        }
     }
 
  #Example 3 (Grab the standard output & standard error of a system call):
 
  use IO::Pipe::Producer;
  $obj = new IO::Pipe::Producer();
  ($stdout_fh,$stderr_fh) =
    $obj->getSystemProducer("echo \"Hello World!\"");


=head1 ABSTRACT

Producer.pm is useful for chaining large data processing subroutines or system calls.  Instead of making each call serially and waiting for a return, you can create a Producer that will continuously generate output that can be immediately processed.  You can even split up input and run subroutines in parallel.  Producer.pm is basically a way to pipe the standard output of a forked subroutine or system call to a file handle in your parent process.

=head1 DESCRIPTION

Producer.pm is a module that provides methods to fork off a subroutine or system call and return handles on the standard output (STDOUT and STDERR).  If you have (for example) a subroutine that processes a very large text file and performs a task on each line, but you need to perform further processing, normally you would have to wait until the subroutine returns to get its output.  If the subroutine prints its output to STDOUT (and STDERR) or you can edit it to do so, you can call it using a Producer so that you can use the returned handle to continuously process each line as it's "produced".  You can chain subroutines together like this by having your subroutine itself create a Producer.  This is similar to using open() to run a system call, except that with this module, you can get a handle on STDERR and use it with subroutines as well.  And by dividing up your input, you can take advantage of multi-core systems and process your data in a parallel fashion.

Note that the handles retuned are open file handles.  It is your job to close them once you are finished with them.

=head1 NOTES

This module was originally written as a simple subroutine that used IO::Pipe.  It adds one method and a helper method (Note: The getSystemProducer method calls getSubroutineProducer).  It functions by opening STDOUT/STDERR as input, which is the basic definition of a pipe.  Those input file handles are what is returned.

=head1 BUGS

No known bugs.  Please report them to I<E<lt>rleach@princeton.eduE<gt>> if you find any.

=head1 SEE ALSO

L<IO::Pipe>
L<IO::Select>

=head1 AUTHOR

Robert William Leach, E<lt>rleach@princeton.eduE<gt>

=head1 COPYRIGHT AND LICENSE

This software and ancillary information (herein called "SOFTWARE") called Producer.pm is made available under the terms described here.  The SOFTWARE has been approved for release with associated LA-CC number LA-CC-05-060.

Unless otherwise indicated, this software has been authored by an employee or employees of the University of California, operator of the Los Alamos National Laboratory under Contract No. W-7405-ENG-36 with the U.S. Department of Energy.  The U.S. government has rights to use, reproduce, and distribute this SOFTWARE.  The public may copy, distribute, prepare derivative works and publicly display this SOFTWARE without charge, provided that this notice and any statement of authorship are reproduced on all copies.  Neither the government nor the university makes any warranty, express or implied, or assumes any liability or responsibility for the use of this SOFTWARE.

If SOFTWARE is modified to produce derivative works, such modified SOFTWARE should be clearly marked, so as not to confuse it with the version available from LANL.


=cut
